src/join-9.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+/// join-9s.sol -- Non-standard token adapters
+
+// Copyright (C) 2018 Rain <rainbreak@riseup.net>
+// Copyright (C) 2018-2020 Maker Ecosystem Growth Holdings, INC.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.12;
+
+import "dss/lib.sol";
+
+interface VatLike {
+    function slip(bytes32, address, int256) external;
+}
+
+interface GemLike {
+    function decimals() external view returns (uint8);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function getFeeFor(uint256) external returns (uint256);
+}
+
+// For a token that has a fee (PAXG)
+
+contract GemJoin9 is LibNote {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth {
+        require(wards[msg.sender] == 1, "GemJoin/not-authorized");
+        _;
+    }
+
+    VatLike public vat;   // CDP Engine
+    bytes32 public ilk;   // Collateral Type
+    GemLike public gem;
+    uint256 public dec;
+    uint256 public live;  // Active Flag
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        gem = GemLike(gem_);
+        dec = gem.decimals();
+    }
+    function cage() external note auth {
+        live = 0;
+    }
+    function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x);
+    }
+    function join(address usr, uint256 wad) external note {
+        require(live == 1, "GemJoin/not-live");
+        require(int256(wad) >= 0, "GemJoin/overflow");
+        vat.slip(ilk, usr, int256(sub(wad, gem.getFeeFor(wad))));
+        require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin/failed-transfer");
```

Comment:
I think itd be worth adding a `uint256 beforeBal = gem.balanceOf(address(this));` before the `transferFrom` and then an `assert`: 
`assert(sub(gem.balanceOf(address(this), beforeBal) == sub(wad, gem.getFeeFor(wad))));` to avoid any surprises down the road. Especially since this is upgradeable. cc @godsflaw

Reply:
Makes sense to me.

Reply:
Was going to come back and mention this after looking it over yesterday. Glad this got added. LGTM.